src/dmd/expressionsem.d

```diff
@@ -5974,6 +6085,32 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             e.e1 = e.e1.expressionSemantic(sc);
         }
         result = e;
+        if (f.vthis2 && !sc.intypeof && sc.func)
+        {
+            if (AggregateDeclaration ad2 = f.isMember2())
+            {
+                Expression te = new ThisExp(e.loc).expressionSemantic(sc);
+                if (te.op != TOK.error)
+                    te = getRightThis(e.loc, sc, ad2, te, f);
+                if (te.op == TOK.error)
+                {
+                    e.error("need `this` of type `%s` to make delegate from function `%s`", ad2.toChars(), f.toChars());
+                    return setError();
+                }
+            }
+            Type tthis2 = Type.tvoidptr.sarrayOf(2);
+            VarDeclaration vthis2 = new VarDeclaration(e.loc, tthis2, Identifier.generateId("__this"), null);
+            vthis2.storage_class |= STC.temp;
+            vthis2.dsymbolSemantic(sc);
+            vthis2.parent = sc.parent;
+            // make it a closure var
+            vthis2.nestedrefs.push(f);
+            sc.func.closureVars.push(vthis2);
```

Comment:
So wait, are you passing around the context pointer as `void*[2]` or `struct{void*[2]}*`?

Bearing in mind that it is an ICE in gdb to have the context pointer anything other than a `struct*`. I have no idea what it will do went the chained 'this' pointer is an array, it certainly won't make debugging easy.

Reply:
what do you mean? it is passed by ref so it will be`void*[2]*`.

Reply:
That's still not a valid this pointer.

Can you really not solve this by adding a new frame to the chain?

Given
```
auto func(alias b)()
{
    ++a;
    ++b;
}
```
The frame can look like:
```
struct aliasframe
{
  struct funcframe
  {
    int a;
  }* this;
  int b;
}
```

The codegen:
```
funcframe __frame1 = { a = 1; }
aliasframe __frame2 = { this = &__frame1, b = 2; }

func(aliasframe* this)
{
  this.this.a++;
  this.b++;
}

func(&__frame2);
assert(__frame1.a == 2);
assert(__frame2.b == 3);
```

Reply:
Or the `alias b` could become a ref parameter, so you only have to worry about one context.

```
funcframe __frame = { a = 1; }

func(funcframe* this, ref int b)
{
  this.a++;
  b++;
}

int b = 2;
func(&__frame, &b);
assert(__frame.a == 2);
assert(b == 3);
```

Reply:
why is it still not a valid this pointer? it's ABI compatible with delegates.

Reply:
```
void main()
{
    int a = 2;
──> void foo() { a++; }
    foo();
}
----
(gdb) p a
Internal error: `this' is not an aggregate
```
This is a dmd bug, which is easily fixable.  What this patch proposes makes it worse.

Reply:
> (gdb) p a
> Internal error: `this' is not an aggregate

I do not understand the source of that problem. is it just typing?
the `void*[2]` could also be casted to `struct {  struct {int a;} f0; struct { int b;} f1;  } this;`
and accessing `a` would be `this.f0.a;` what wrong with this?